### PR TITLE
Adding Python 3.7 Support: async is now a reserved keyword

### PR DESCRIPTION
**Pull request by jvasallo**
- renamed references to async to c7n_async to avoid confusion

I am open to a better name but this was the first one that came to mind.

### CODE REVIEW

**Comment by kapilt** [comment]
Thanks for the pull request!  If you don't mind could you sign the CLA at the bottom of the README? Direct link here: https://docs.google.com/forms/d/e/1FAIpQLSfwtl1s6KmpLhCY6CjiY8nFZshDwf_wrmNYx1ahpsNFXXmHKw/viewform

The s/async/c7n_async changes look good minus the action.py module, where its actually a user configuration string which we'll need to maintain for compatibility, for that case we'll need to define the schema manually instead of using the type_schema helper.

**Comment by jvasallo** [comment]
Hey @kaplit I filled out the CLA earlier (2-4 hours ago), maybe it didn't process yet?

I will review action.py soon! Thanks for the feedback!

**Comment by kapilt** [comment]
@jvasallo thanks and sorry about that, it appears we're having some collection issues on that form, i'm trying to track it down.

**Comment by kapilt** [comment]
historical context.. I was trying to track down our use of ruamel through the code base to audit compatibility, but was trying to remember why we have it at all. It was something @chadwhitacre did to help py2/3 compatibility specifically in tools/c7n_mailer for lambda support (normal custodian policies don't have a runtime dep on yaml in lambda) as pyyaml had issues deploying lambdas across language versions (ie. 3.6 cli deploying a 2.7 lambda or vice versa) due to the internal details of the package. everywhere else we use pyyaml.

I installed the latest ruamel.yaml==0.15.42 and it looks like its compatible with the mailer usage, so upgrading that version pin to 0.15.42 sounds good for py3.7 compatibility.

File: .travis.yml [hunk 1]
```diff
@@ -1,14 +1,21 @@
...
+    include:
+        - python: 2.7
+          dist: trusty
+          sudo: false
+        - python: 3.6
+          dist: trusty
+          sudo: false
+        - python: 3.7
+          dist: xenial
+          sudo: true
```
Comment (kapilt) [inline .travis.yml]: do we need the sudo?

**Comment by kapilt** [comment]
you'll need one more change to restrict the 3.7 ci execution to the test runs in. tools/dev/citoxenv.py you'll need another clause for 3.7 that looks like the one for 3.6. This will restrict the 3.7 matrix build to just run the tests instead of also doing the sphinx/coverage/lint runs.

File: .travis.yml (hunk 1 above)
Comment (jvasallo) [reply .travis.yml]: Yeah. I am a bit in shock on that too. I think 3.7 is still being worked into Travis. 
https://github.com/travis-ci/travis-ci/issues/9815#issuecomment-401756442

Sounds like someone got it to work on xenial without sudo so I will try that now.
https://github.com/travis-ci/travis-ci/issues/9815#issuecomment-403142521

Sounds like its an issue of a missing openssl library on trusty:
https://docs.travis-ci.com/user/languages/python/#Development-releases-support

Travis Issue:
https://github.com/travis-ci/travis-ci/issues/9069

File: c7n/actions.py [hunk 2]
```diff
@@ -359,7 +363,7 @@ def process(self, resources, event=None):
         if self.data.get('qualifier'):
             params['Qualifier'] = self.data['Qualifier']
 
-        if self.data.get('async', True):
+        if self.data.get('c7n_async', True):
```
[the remark this answers is not in the document]
Comment (jvasallo) [reply c7n/actions.py]: Good catch! 👍Thanks!

File: tools/c7n_mailer/requirements.txt [hunk 3]
```diff
@@ -10,7 +10,7 @@ Markup
 MarkupSafe==1.0
 python-dateutil>=2.5.2
 ldap3
-ruamel.yaml<0.15 # rq.filter: <0.15
+ruamel.yaml==0.15.42 # rq.filter: <0.15
```
[the remark this answers is not in the document]
Comment (jvasallo) [reply tools/c7n_mailer/requirements.txt]: Dropped! Thanks for the info.

File: c7n/actions.py [hunk 4]
```diff
@@ -336,13 +336,17 @@ class LambdaInvoke(EventAction):
...
-    schema = utils.type_schema(
-        'invoke-lambda',
-        function={'type': 'string'},
-        async={'type': 'boolean'},
-        qualifier={'type': 'string'},
-        batch_size={'type': 'integer'},
-        required=('function',))
+    schema = {
+        'type': 'object',
+        'required': ['function'],
```
[the remark this answers is not in the document]
Comment (jvasallo) [reply c7n/actions.py]: Added `type` to required. Missed that note in the `type_schema` def. Thanks!

**Comment by jvasallo** [comment]
Fixed issue with py37 env. Thanks for that info as well!

**Comment by kapilt** [comment]
I think we're down to the last leg, thanks again for working on this. So the last ci failure is a bit more interesting, and may require some additional changes, its specifically in regards to handling namespace packages in python 3.7 which is either a behavior change in 3.7 or a test issue.

scanning through the 3.7 changeling this part shows a relevant bit

`
Fixed some consistency problems with namespace package module attributes. Namespace module objects now have an __file__ that is set to None (previously unset), and their __spec__.origin is also set to None (previously the string "namespace"). See bpo-32305. Also, the namespace module object’s __spec__.loader is set to the same value as __loader__ (previously, the former was set to None). See bpo-32303.`

re last ci failure
````
Traceback (most recent call last):
  File "/home/travis/build/capitalone/cloud-custodian/tests/test_mu.py", line 825, in test_handles_namespace_packages
    self.assertTrue(filenames[-1].endswith("-nspkg.pth"))
  File "/opt/python/3.7.0/lib/python3.7/unittest/case.py", line 692, in assertTrue
    raise self.failureException(msg)
AssertionError: False is not true
```

so looking at the code I'm guessing the fix is in c7n/mu.py line 91 the hasattr check that guards namespace handling, since there is a `__file__` attribute now (with `None` value) and that guard condition needs to be updated to also allow a None attribute. This change should do it.

`if getattr(module, '__file__', None) is None:`

**Comment by jvasallo** [comment]
Testing the above fix locally for py27, py36, and py37. So far so good. I must have missed the py37 failure last night apologies! Commit with fix coming soon. Thanks for the help, learning a bit here too 👍

**Pull request merged**